ca_gazebo/src/color_sensor_plugin.cpp

```diff
@@ -0,0 +1,131 @@
+        << "Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
+      return;
+    }
+
+    CameraPlugin::Load(_parent, _sdf);
+    // copying from CameraPlugin into GazeboRosCameraUtils
+    this->parentSensor_ = this->parentSensor;
+    this->width_ = this->width;
+    this->height_ = this->height;
+    this->depth_ = this->depth;
+    this->format_ = this->format;
+    this->camera_ = this->camera;
+    this->_rvalue_max = _sdf->GetElement("rvalue_max")->Get<double>();
+    this->_rvalue_min = _sdf->GetElement("rvalue_min")->Get<double>();
+    this->_gvalue_max = _sdf->GetElement("gvalue_max")->Get<double>();
+    this->_gvalue_min = _sdf->GetElement("gvalue_min")->Get<double>();
+    this->_bvalue_max = _sdf->GetElement("bvalue_max")->Get<double>();
+    this->_bvalue_min = _sdf->GetElement("bvalue_min")->Get<double>();
+    this->_publisher_name = _sdf->Get<std::string>("PublisherName");
+     _sensorPublisher = _nh.advertise<std_msgs::Bool>(this->_publisher_name, 1);
+     this->_factor_color_per_image=_sdf->GetElement("FactorColorperImage")->Get<double>();
+     this->parentSensor_->SetActive(true);
+
+    GazeboRosCameraUtils::Load(_parent, _sdf);
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  // Update the controller
+  void GazeboRosColor::OnNewFrame(const unsigned char *_image,
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+    static int seq=0;
+
+    int R,G,B;
+
+    double _desired_color_in_pixel=0, _pixel_count=0;
+
+    this->sensor_update_time_ = this->parentSensor_->LastUpdateTime();
+
+    if (!this->parentSensor->IsActive())
+    {
+      if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run once after activated
+        this->parentSensor->SetActive(true);
+    }
+    else
+    {
+      if ((*this->image_connect_count_) > 0)
+      {
+        common::Time cur_time = this->world_->SimTime();
+        if (cur_time - this->last_update_time_ >= this->update_period_)
+        {
+          this->PutCameraData(_image);
+          this->PublishCameraInfo();
+          this->last_update_time_ = cur_time;
+
+          bool _output;
+
+          for (int i=0; i<_width*_height*3 ;i=i+3)
+          {
+              R=_image[i] ;
+              G=_image[i + 1];
+              B=_image[i + 2];
+              _pixel_count++;
+              if (R<=this->_rvalue_max and R>this->_rvalue_min and G<=this->_gvalue_max and G>this->_gvalue_min and B<=this->_bvalue_max and B>this->_bvalue_min)
+              {
+                _desired_color_in_pixel++;
+              }
+            }
+
+          if(_desired_color_in_pixel/_pixel_count > this->_factor_color_per_image)
+          {
+            _output=true;
+          }
+          else
+          {
+            _output=false;
+          }
+        _sensorPublisher.publish(_output); 
```

Comment:
Apply [this](http://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Intraprocess_Publishing) workaround.